saas/control_app/src/TenantManagement/functions/provisioning.py

```diff
@@ -0,0 +1,108 @@
+import json
+import re
+# import sys
+
+# sys.path.append(os.path.join(os.path.dirname(__file__), 'functions'))
+from TenantManagement.functions.shared.db_repository import AwsDbRepository
+from TenantManagement.functions.shared.db_repository import (
+    DbRepository, TenantIdAlreadyInUseException, EmailAlreadyInUseException)
+from TenantManagement.functions.shared import tenant_utils
+
+from TenantManagement.functions.shared.tenant import Tenant
+from TenantManagement.functions.shared.tenant_utils import create_random_password
+from TenantManagement.functions.shared.tenant import DeploymentState
+from TenantManagement.functions.shared.infra_repository import (InfraRepository, BotoInfraRepository)
+
+region = os.environ['REGION'] if 'REGION' in os.environ else "ddblocal"
+db_table = os.environ['DB_TABLE'] if 'DB_TABLE' in os.environ else "saas"
+db_endpoint = os.environ['DB_ENDPOINT'] if 'DB_ENDPOINT' in os.environ else None
+
+EMAIL_REGEX = re.compile(r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$")
+
+
+class EventException(Exception):
+    pass
+
+
+def assert_event_key(event: dict, *keys: str):
+    _event = event
+    path = keys[0]
+    result = None
+    for index, key in enumerate(keys, start=1):
+        if index == len(keys):
+            if key in _event:
+                result = _event[key]
+            else:
+                raise EventException(f"event does not contain \"{path}\"")
+
+        if key in _event:
+            _event = _event[key]
+            path = f"{path}.{key}"
+        else:
+            raise EventException(f"event does not contain \"{path}\"")
+
+    return result
+
+def build_response(message: str, status_code: int):
+    return {
+        "statusCode": status_code,
+        "body": json.dumps({"message": message}),
+        "headers": {
+            'Content-Type': 'application/json',
+        }
+    }
+
+
+def handle_signup(email: str, db_repo: DbRepository, infra_repo: InfraRepository) -> dict:
+    if not EMAIL_REGEX.match(email):
+        return build_response("Email address is not valid", 400)
+
+    tenant = Tenant(email=email, id=tenant_utils.create_id(), auth_keys={}, plan_keys={
+    }, deployment_state=DeploymentState.NEEDS_UPDATE).update_auth_defaults(create_random_password())
+    try:
+        tenant = db_repo.save_new_tenant(tenant)
+    except TenantIdAlreadyInUseException:
+        # Retry to create another tenantId in case of a single collision, in this case we just let the requester know that they need to try that again.
+        logging.exception(
+            "Couldn't generate a unique id, abort the process.")
+        return build_response("Can not create tenant, try again later. ", status_code=500)
+    except EmailAlreadyInUseException:
+        # Actually this message should be "Email already in use, but this could be a spoofing mechanism to detect use emails from us. TODO intoduce a captcha, mechanism"
+        return build_response("Can not create tenant", status_code=409)
```

Comment:
in that case we can also return 400 to not reveal any information
```suggestion
        return build_response("Can not create tenant", status_code=400)
```